stackrox/stackrox
# Write frontend config and telemetry proto and rewrite services in TypeScript

## Description

* Increase type safety with compile-time checks.
* Simplify maintenance of consistent backend and frontend response types.

### Changed files

1. Edit to import from frontend proto files
    * Containers/SystemConfig/ConfigBannerDetailWidget.tsx
    * Containers/SystemConfig/ConfigDataRetentionDetailWidget.tsx
    * Containers/SystemConfig/ConfigLoginDetailWidget.tsx
    * Containers/SystemConfig/ConfigTelemetryDetailWidget.tsx
    * Containers/SystemConfig/Details.tsx
    * Containers/SystemConfig/Page.tsx
    * Containers/SystemConfig/SystemConfigForm/SystemConfigForm.tsx

2. Delete Containers/SystemConfig/SystemConfigTypes.ts

3. Rewrite as TypeScript
    * services/SystemConfigService.js
    * services/TelemetryService.js

4. Add frontend proto files
    * types/config.proto.ts
    * types/telemetry.proto.ts

### Residue

1. Move **Save** and **Cancel** buttons and Formik state into `SystemConfigForm` component.
    * Among other reasons, there is a racing condition because `initialValues` depends on the responses to different requests.
    * Also replace `{ publicConfig, privateConfig, telemetryConfig }` with `{ systemConfig, telemetryConfig }` in values, because payload to update system configuration incorrectly includes telemetry configuration.
2. Reduce redux store to `publicConfig` for `LoginNotice` and `MainPage` components.
3. Reduce sagas/systemConfig.js to initial get for `loginPath` and `mainPath` but move update to SystemConfig container.
4. Move requests from sagas/telemetrySagas.js to SystemConfig container.
5. Make sure integration tests do not make changes which prevent multiple runs in local deployment.
6. Add relevant intercept and wait to helpers for system configuration integration tests.

## Checklist

## Testing Performed

### manual testing

1. `yarn build` in ui
2. `yarn start` in ui

    * view initial has `publicConfig: null` in response from /v1/config request
        ![SystemConfiguration-view](https://user-images.githubusercontent.com/11862657/168679016-35737715-4d30-4858-a56b-9b6fe7c5e962.png)

    * form initial has `publicConfig: null` in response from /v1/config request
        ![SystemConfiguration-edit](https://user-images.githubusercontent.com/11862657/168679046-491b7f57-c5f7-4262-bfe5-c9bf93aa0656.png)

    * main dashboard has `{loginNotice: null, header: null, footer: null}` in response from /v1/config/public request
        ![public-null-main](https://user-images.githubusercontent.com/11862657/168680035-7f7cd83d-979d-4503-bee2-f97499162e53.png)

    * `SystemConfigPage` has a racing condition because `initialValues` depends on the responses to different requests: notice `telemetryConfig: {}` and then with value from response to request
        ![Page-initialValues](https://user-images.githubusercontent.com/11862657/168680017-94a32e6f-b220-4c2c-a418-0ae0abbd5156.png)

    * `onSubmit` leaks `telemetryConfig` into system configuration update and notice that, because of racing condition, it has initial `{}` value before response to request
        ![loginNotice](https://user-images.githubusercontent.com/11862657/168680553-81e2a449-3697-4708-8fa3-2c4c9b0a419b.png)

### integration testing

1. `yarn cypress-open` in ui/apps/platform
    * systemconfig.test.js
